### PR TITLE
Xamarin Forms Gtk# Backend - Platform Project

### PR DESCRIPTION
**Pull request by jsuarezruiz**
### Description of Change ###

**Preview** of Xamarin.Forms.GTK project and Xamarin.Forms.Maps.GTK project. 

**NOTE:** This PR only includes the **Xamarin.Forms.Platform.GTK** project. Other PRs will include maps, CoreGallery and other points.

### Bugs Fixed ###

- None.

### API Changes ###

Added:

- Custom Widgets, helpers, extensions, etc.
- ActivityIndicatorRenderer (Custom Widget that makes use of Cairo.).
- BoxViewRenderer (Custom Widget that makes use of Cairo.).
- ButtonRenderer (Gtk.Button).
- CarouselPageRenderer (Custom Widget).
- Cells (EntryCell, ImageCell, SwitchCell, TextCell and ViewCell).
- DatePickerRenderer (Custom Widget using Gtk.Window + Gtk.Calendar).
- EditorRenderer (Gtk.TextView).
- EntryRenderer (FormsTextBox).
- FrameRenderer (Gtk.Frame).
- ImageRenderer (Gtk.Image).
- LabelRenderer (Gtk.Label).
- LayoutRenderer (Gtk.Container).
- ListViewRenderer (Custom Widget).
- MasterDetailPageRenderer (Custom Widget).
- NavigationPageRenderer (Custom Widget).
- OpenGLViewRender (OpenGL Widget that uses OpenTK).
- PageRenderer (Gtk.Container).
- PickerRenderer (Gtk.ComboBox).
- ProgressBarRenderer (Gtk.ProgressBar).
- ScrollViewRenderer (Gtk.ScrolledWindow).
- SearchBarRenderer (Custom Widget).
- SliderRenderer (Gtk.HScale).
- StepperRenderer (Gtk.SpinButton).
- SwitchRenderer (Gtk.CheckButton).
- TableViewRenderer (Custom Widget).
- TimePickerRenderer (Custom Widget using Gtk.Window + 3 Gtk.SpinButton).
- WebViewRenderer (Custom Widget that uses webkit-sharp).

### Behavioral Changes ###

You can create applications for **Linux**, MacOS and Windows with Xamarin.Forms using the GTK# backend.

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [ ] Rebased on top of master at time of PR
- [ ] Changes adhere to coding standard
- [ ] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.Platform.GTK/Animations/BaseAnimation.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
+﻿using System;
+using System.Threading.Tasks;
+using System.Timers;
+
+namespace Xamarin.Forms.Platform.GTK.Animations
+{
+    public abstract class BaseAnimation
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.GTK/Animations/BaseAnimation.cs]: Why do we want this public ?

File: Xamarin.Forms.Core/Device.cs [hunk 2]
```diff
@@ -16,8 +16,9 @@ public static class Device
 		public const string UWP = "UWP";
 		public const string WinRT = "WinRT";
 		public const string macOS = "macOS";
+        public const string GTK = "GTK";
 
-		[EditorBrowsable(EditorBrowsableState.Never)]
+        [EditorBrowsable(EditorBrowsableState.Never)]
```
Comment (rmarinho) [inline Xamarin.Forms.Core/Device.cs]: seems all PR has problems with code guidelines,we use spaces not  tabs, we also don't use private

File: Xamarin.Forms.Platform.GTK/Cells/CellBase.cs [hunk 3]
```diff
@@ -0,0 +1,150 @@
+﻿using Gtk;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+
+namespace Xamarin.Forms.Platform.GTK.Cells
+{
+    public abstract class CellBase : EventBox
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.GTK/Cells/CellBase.cs]: We don't want to make lot's of thigns public, i maybe recomend have a Internals namespace at least

File: Xamarin.Forms.Platform.GTK/Animations/BaseAnimation.cs (hunk 1 above)
Comment (jsuarezruiz) [reply Xamarin.Forms.Platform.GTK/Animations/BaseAnimation.cs]: It is used in gtk widgets to perform certain animations. For example, with the MasterDetailPage. I do not think it's necessary to be public having all the animations API in Xamarin.Forms. Changed!.

File: Xamarin.Forms.Platform.GTK/Cells/CellBase.cs (hunk 3 above)
Comment (jsuarezruiz) [reply Xamarin.Forms.Platform.GTK/Cells/CellBase.cs]: Yeah, good one. Changed native cell controls to become internal classes.

Comment (laarmen) [reply Xamarin.Forms.Platform.GTK/Cells/CellBase.cs]: Out of curiosity, is there a reason to keep so many things internal ? From the lib client point of view, it often makes it painful to customize a renderer behaviour

**Comment by rmarinho** [comment]
this is merged!

**Pull request closed without merging**